market-generator/src/main/java/com/market/banica/generator/service/grpc/MarketService.java

```diff
@@ -34,11 +54,110 @@ public void subscribeForItem(MarketDataRequest request, StreamObserver<TickRespo
         }
     }
 
+    @Override
+    public void buyProduct(ProductBuySellRequest request, StreamObserver<BuySellProductResponse> responseObserver) {
+
+        cleanPendingOrdersCollection(request);
+
+        BuySellProductResponse buySellProductResponse = BuySellProductResponse.newBuilder().setMessage(String.format("Item with name %s was successfully bought from %s market.", request.getItemName(), request.getMarketName())).build();
+
+        responseObserver.onNext(buySellProductResponse);
+        responseObserver.onCompleted();
+
+    }
+
+    @Override
+    public void returnPendingProduct(ProductBuySellRequest request, StreamObserver<BuySellProductResponse> responseObserver) {
+
+        cleanPendingOrdersCollection(request);
+        marketState.addGoodToState(request.getItemName(), request.getItemPrice(), request.getItemQuantity(), request.getTimestamp());
+
+        BuySellProductResponse buySellProductResponse = BuySellProductResponse.newBuilder().setMessage(String.format("Item with name %s was successfully returned to market.", request.getItemName())).build();
+
+        responseObserver.onNext(buySellProductResponse);
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void checkAvailability(ProductBuySellRequest request, StreamObserver<AvailabilityResponse> responseObserver) {
+        boolean isAvailable = false;
+        MarketTick marketTick = new MarketTick();
+
+        String productName = request.getItemName();
+        double productPrice = request.getItemPrice();
+        long productQuantity = request.getItemQuantity();
+        String marketName = request.getMarketName();
+
+        try {
```

Comment:
Wouldn't this be better if it is implemented with an "if" and not throwing exception as this is a case that is not out of the ordinary when someone buys a product before us.